dogechain-lab/dogechain
# Feat faster bulk syncing

# Description

The syncing implementation is very highly coupling with IBFT consensus, which is not well designed and buggy all the time. We migrate the Polygon Edge `/syncer/0.2` protocol to current protocol version, and make it backward compatible.

Important changes:

* Separate `Syncer` from `IBFT` consensus.
* Update `syncer` protocol for backward compatibility.
* Use gRPC stream instead of unary call to speed up sync.
* Stop broadcasting blocks during consensus rounds.
* Stop broadcasting blocks during bulk sync until explicitly set.
* Remove common ancestor queries.
* Only trigger sync when new topic `/dogechain/syncer/status/0.1` is received.

# Changes include


# Breaking changes

An old version node can only initially sync up to the height it is querying from peers, otherwise it will stop catching up until the new version updates, or some generous peers enable the broadcast block feature.

## Testing


### Manual tests

#### test case 1

* Set up a 4 validators and 2 full nodes network.
* Send some transactions.
* Check the status and syncing status.

It should be OK and get `{"jsonrpc":"2.0","id":0,"result":false}` sync status.

#### test case 2

* Set up a 4 validators network with `block-broadcast` enable.
* Send some transactions and wait enough blocks to observe sync status (e.g. 10-20 minutes or so).
* Set up two full nodes with different version (PR branch and target branch).
* Check the two full nodes' status and syncing status.

It should be OK and get something continuously updated like `{"jsonrpc":"2.0","id":0,"result":{"type":"bulk-sync","startingBlock":"0x333b85","currentBlock":"0x334347","highestBlock":"0x331f7c"}}` sync status.
